### ISSUE TITLE
should we generate typscript type files by default?

### ISSUE THREAD

**Issue by ashleygwilliams** [ashleygwilliams/wasm-pack#114]
a question came up in solving issue #107 on PR #109: should we generate typescript type files by default?

i have requested that the PR land behind a flag but i'm curious to hear the arguments for why it should be default! this is the issue for that convo :)

**Comment by kwonoj** [comment]
As shared in https://github.com/ashleygwilliams/wasm-pack/pull/109#issuecomment-384464904 I vote for generate by default, opt out via cmd args when invoke wasm-pack. Lot of tool understands type definition and provides additional info in js code even, so it benefit js/ts users both.

**Comment by Pauan** [comment]
I vote for yes as well: there is no downside to generating `.d.ts`

**Comment by kwonoj** [comment]
for pkg size - 
1. Installation size (under `node_modules`) will be increased, yes, but mostly really small size. due to nature of type definition, say i.e you have javascript fn
```
function doSomething(args) {
//really long long long implementation
}
```

it'd corresponding type definition simply becomes single-line
```
function doSomething(args: any): any;
```

Course if it have some interfaces, or enums / dependent types it'll be increased more, but it's plain text files notably small size.

This is one of real-world comparision between impl to type definition from codebase I mostly work with - 

actual impl
https://cdn.jsdelivr.net/npm/rxjs@6.0.0/internal/operators/audit.js
TypeDef
https://cdn.jsdelivr.net/npm/rxjs@6.0.0/internal/operators/audit.d.ts

you can see differences.

2. When it comes to production (for browsers), `d.ts` shouldn't be matter as actual js bundle never includes any type definitions.

**Comment by kwonoj** [comment]
For npm pkgs deal with types - if module author intends to support types by default, it is common to ship type definition along with packages. Most cases it depends on author's preferences, between either 1. would like to support types (either flow, or ts) 2. or doesn't have preferences for it and let 3rd party deals with it. Size usually isn't concern compare to actual codebase size which is real concern in production.

**Comment by ashleygwilliams** [comment]
@kwonoj the point about bundlers is a good one. ok. i'm coming around to this. perhaps we *can* land #109 without a flag... curious about what @yoshuawuyts @jamiebuilds might think

**Comment by Pauan** [comment]
The [official TypeScript recommendation](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) is to publish `.d.ts` along with `.js` inside of your npm package (assuming your package was designed for TypeScript in the first place).

The other alternative is the `@types` organization, which is intended more for publishing `.d.ts` for third-party packages which are pure-JS (and not intended as a TypeScript package by the author). It's a massive pain to use, though: you have to make a pull request, get it accepted, etc.

**Comment by robertohuertasm** [comment]
FWIW, I totally subscribe @kwonoj 's point of view. I can only see benefits for the consumers of the future packages (no matter they use js or ts). 

In my opinion, as an author you should try to provide the widest support possible, and as a heavy TypeScript user, I really, really appreciate when packages are delivered with a synchronized `d.ts` file. 

As @Pauan already said, it's a hard and painful process to have to depend on 3rd party typings that usually don't follow the same rythm in terms of updates and, hence, tend to be utterly unsynchronized.

**Comment by ashleygwilliams** [comment]
ok so, having slept on this here's the deciding question i think:

is there any case in which a crate written in rust and then compiled with wasm-bindgen would NOT benefit from a types file?

@Pauan wrote "assuming your package was designed for TypeScript in the first place" <-- most people don't design a package for type script, tho writing a package with Rust and compiling to wasm might take the need to design for type script out of the question?

assuming there is ZERO effort required on the authors part to potentially have the types be useful for the package, let's do it by default. otherwise let's not. does that make sense?

@alexcrichton i'm curious- why did you make the typescript definitions opt in? we're discussing making them the default here and i think it'd be a little weird for bindgen and pack to have different defaults...

**Comment by kwonoj** [comment]
> compiled with wasm-bindgen would NOT benefit from a types file?

The only case I can currently think of is if types generated from bindgen does not corretly match to signature of original rust fn for some reason, gives incorrect signature to consumers.

**Comment by alexcrichton** [comment]
@ashleygwilliams some good points! I think it's actually a historical accident that `wasm-bindgen` doesn't emit *.d.ts files by default. Originally it *only* emitted typescript and then got an option to emit JS instead. Later I realied that *.d.ts existed and wanted to switch to JS by default so I swapped the two. I don't think I ever stopped to realize that *.d.ts is harmless and would be good to generate by default!

tl;dr; I'd be totally fine generating *.d.ts by default, and I agree wasm-pack and wasm-bindgen shoud have the same defaults

**Comment by alexcrichton** [comment]
I've opened https://github.com/rustwasm/wasm-bindgen/pull/173 to change wasm-bindgen's default, which can probably land in tandem with the corresponding wasm-pack PR

**Comment by jamiebuilds** [comment]
Could we make it generate both TypeScript and Flow definition files? They are very similar definition formats and it would work basically the same way with a `.js.flow` file next to the `.js` file

**Comment by mgattozzi** [comment]
@jamiebuilds this would require wasm-bindgen to support them first at least. Otherwise wasm-pack can't even generate them! I think that's up to @alexcrichton and whether he wants wasm-pack to handle that

**Comment by alexcrichton** [comment]
@jamiebuilds sounds plausible to me! Want to open an issue at wasm-bindgen?

**Comment by konstin** [comment]
What about using something like [flowgen](https://github.com/joarwilk/flowgen)? The actual conversion could be done in wasm-pack, while the CI in wasm-bindgen checks that the emitted typescript is understood by flowgen.

**Comment by Pauan** [comment]
@sendilkumarn There's only two major JS type systems: TypeScript and Flow. [And TypeScript is *by far* the most common](https://trends.google.com/trends/explore?date=all&q=typescript,flow%20js), it is the de-facto standard.

And if the cost of maintaining flow types is simply running flowgen, that doesn't sound like a big maintenance burden.

Also, if we provide the feature *at all* (even behind a flag), it has the *same* maintenance burden, since we have to make sure that it is correct and supported (because the user *might* use the flag).

As for using them "when needed", we're talking about creating libraries, and it is not the *library author* that needs the types, it is the *library consumer*. And obviously the library author cannot predict whether all possible library consumers will need the types or not.

**Comment by sendilkumarn** [comment]
>  if wasm-pack provides the feature at all (even behind a flag), it has the same maintenance burden,

Yeah absolutely agree to this. But for the flag I (personally) feel it is okay to be opinionated and support them and choose not to support others.

> that needs the types, it is the library consumer

AFAICT when your library grows and even as a library author you might need type system. 

>  And obviously the library author cannot predict whether all possible library consumers will need the types or not.

Exactly that is what is running on my mind.

But I think it is a thin line here. There is no second thought about the values that type system provides, (in fact I have used them both and enjoyed their benefits). And also provided that `wasm-bindgen` now supports it by default. I think it is good to let it be like that 👍

**Comment by Pauan** [comment]
> AFAICT when your library grows and even as a library author you might need type system.

The library author is writing their code in Rust, which is already fully statically typed. What we are talking about is a system which takes your Rust types and *automatically* translates them into TypeScript types, so that people using your Rust crate as a library will benefit. The types will always be in sync with the Rust types.

**Comment by yoshuawuyts** [comment]
Chiming in a day late here, but yeah all in favor of adding TS type files out of the box. Not seeing any downsides and tons (!) of upsides!

**Comment by ashleygwilliams** [comment]
hey ya'll! it's decided :) we're gonna add TS type files out of the box and we'll see how to do it for flow as well. thanks!

**Comment by mgattozzi** [comment]
We should consider this issue closed then. Unless you want to wait for #109 to land

**Comment by ashleygwilliams** [comment]
yeah i'm waiting for #109 to land.

**Comment by chriskrycho** [comment]
One note to a comment made [up-thread](https://github.com/ashleygwilliams/wasm-pack/issues/114#issuecomment-385034871)—simply informational, since shipping types is already a thing that's happening:

> If an ES6 project imports the compiled Rust crate, and it doesn't make use of the TypeScript types at all, then obviously it won't benefit.

This is true in a *very* strict sense, if you're using something like `ed` as your editor. Increasingly, however, many editors and IDEs will actually use TS definitions to make the _JS_ experience better when interacting with a package that has type definitions. (Depending on configuration, some editors may also do the same with Flow types; I haven't checked recently.) I know that at last VS Code and the JetBrains IDEs do this; I *believe* the Vim TS plugin may support doing the same from JS files (but I wouldn't quote me on that!).

**Comment by mgattozzi** [comment]
@chriskrycho oh that's real neat! :D

**Comment by ashleygwilliams** [comment]
solved by #109
